crates/uv-traits/src/lib.rs

```diff
@@ -288,6 +290,112 @@ impl NoBuild {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct ConfigSettingEntry {
+    /// The key of the setting. For example, given `key=value`, this would be `key`.
+    key: String,
+    /// The value of the setting. For example, given `key=value`, this would be `value`.
+    value: String,
+}
+
+impl FromStr for ConfigSettingEntry {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some((key, value)) = s.split_once('=') else {
+            return Err(anyhow::anyhow!(
+                "Invalid config setting: {s} (expected `KEY=VALUE`)"
+            ));
+        };
+        Ok(Self {
+            key: key.trim().to_string(),
+            value: value.trim().to_string(),
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum ConfigSettingValue {
+    /// The value consists of a single string.
+    String(String),
+    /// The value consists of a list of strings.
+    List(Vec<String>),
+}
+
+/// Settings to pass to a PEP 517 build backend, structured as a map from (string) key to string or
+/// list of strings.
+///
+/// See: <https://peps.python.org/pep-0517/#config-settings>
+#[derive(Debug, Default, Clone)]
+pub struct ConfigSettings(BTreeMap<String, ConfigSettingValue>);
+
+impl FromIterator<ConfigSettingEntry> for ConfigSettings {
+    fn from_iter<T: IntoIterator<Item = ConfigSettingEntry>>(iter: T) -> Self {
+        let mut config = BTreeMap::default();
+        for entry in iter {
+            match config.entry(entry.key) {
+                Entry::Vacant(vacant) => {
+                    vacant.insert(ConfigSettingValue::String(entry.value));
+                }
+                Entry::Occupied(mut occupied) => match occupied.get_mut() {
+                    ConfigSettingValue::String(existing) => {
+                        let existing = existing.clone();
+                        occupied.insert(ConfigSettingValue::List(vec![existing, entry.value]));
+                    }
+                    ConfigSettingValue::List(existing) => {
+                        existing.push(entry.value);
+                    }
+                },
+            }
+        }
+        Self(config)
+    }
+}
+
+impl ConfigSettings {
+    /// Convert the settings to a string that can be passed directly to a PEP 517 build backend.
+    pub fn escape_for_python(&self) -> String {
```

Comment:
Since it's all strings that should be the mutual subset of json and python, can we have the escaping done by serde?